usr/src/uts/common/fs/zfs/zil.c

```diff
@@ -1512,142 +1835,787 @@ zil_async_to_sync(zilog_t *zilog, uint64_t foid)
 	}
 }
 
+/*
+ * This function will prune commit itxs that are at the head of the
+ * commit list (it won't prune past the first non-commit itx), and
+ * either: a) attach them to the last lwb that's still pending
+ * completion, or b) skip them altogether.
+ *
+ * This is used as a performance optimization to prevent commit itxs
+ * from generating new lwbs when it's unnecessary to do so.
+ */
 static void
-zil_commit_writer(zilog_t *zilog)
+zil_prune_commit_list(zilog_t *zilog)
 {
-	uint64_t txg;
 	itx_t *itx;
-	lwb_t *lwb;
-	spa_t *spa = zilog->zl_spa;
-	int error = 0;
 
-	ASSERT(zilog->zl_root_zio == NULL);
+	ASSERT(MUTEX_HELD(&zilog->zl_writer_lock));
```

Comment:
I think the naming of this new mutex is problematic, and this new function demonstrates it most clearly. Here you've asserted that a mutex named \`zl_writer_lock' is held; and yet further down, the code in a loop repeatedly locks/unlocks a mutex named \`zl_lock'.

Reply:
I'm not sure I understand. The `zl_writer_lock` and `zl_lock` are used for different purposes.

The `zl_writer_lock` is used solely to ensure only a single thread creates and issues an "lwb" at any given time. Thus, we acquire it in only a couple places; in `zil_commit_writer` and `zil_commit_waiter_timeout`. The `zl_lock` is acquired in many different places, and it's OK to acquire the `zl_lock` after already holding the `zl_writer_lock`. The locking you see here is by design.

With that said, I'm not particularly happy with the naming scheme. The `zl_lock` name was in place prior to this patch, and I couldn't think of a descriptive name that would be any better at describing what it is protecting.

If the comment is more about how the locking in the ZIL is tricky and hard to understand, then I definitely agree. I tried to make it better (not sure if I succeeded in that), but I purposely did not try to change much of the locking surrounding the `zl_lock` and what it protects.

Reply:
I am not raising an issue with the logic—just with the contra-intuition introduced by your naming \`zl_writer_lock'. Especially w.r.t. an object named \`zl_lock'—an object named \`zl_writer_lock' is intuitively related to the former (intuitively, but not actually) and intuitively provides a higher isolation (but again intuition is incorrect because they're not related, and as you say, the former can be acquired after the latter).

I suggest to leave out "write" and name it \`zl_issuer_lock'.

Reply:
I see, thanks for the clarification. I'm OK changing the name of the lock to what you suggest; whatever makes the logic easier to read and understand is worth doing (nomenclature matters).

Would you be OK if I did that in a follow up patch? I've already gotten approval to land this to illumos as-is, so I'm hesitant to change it just prior to landing.

Reply:
It's ok to make a nominal final change as long as you refer to it in the last message to advocates.

E.g., I've sent a final RTI with a note such as "One last change since review, fixed cstyle problem revealed by full-lint build."

So a nominal rename (with a quick, validating, incremental build) should be fine. Also, I do believe advocates will want a one-commit patch submitted, so you'll want to squash as well, no?

Reply:
Yes. I squash them to a single commit before pushing to illumos and openzfs. Leaving them split into multiple commits prior to merging can make it easier for folks to see the individual revisions for a change more easily (mostly to workaround a deficiency in GitHub's UX when PRs are rebased).

Reply:
Oh, I see you already provided Dan with a squashed patch. I guess I was bit too late!